### PR TITLE
Enter the abstraction period for the requirements for returns page

### PR DESCRIPTION
**Pull request by rvsiyad**
https://eaflood.atlassian.net/browse/WATER-4296

This is PR is focused on adding functionality to page 3 of 11 in the return requirement pages.

The pages were previously setup as part of a separate PR to create the user journey.

### CODE REVIEW

File: app/validators/return-requirements/abstraction-period.validator.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+ */
+
+const Joi = require('joi').extend(require('@joi/date'))
+
+const { leftPadZeroes } = require('../../presenters/base.presenter.js')
+
+/**
+ * Validates data submitted for the `/return-requirements/{sessionId}/abstraction-period` page
+ *
+ * When setting up a requirement, users must specify an abstraction period for the return requirement. Users must input a valid from and to
```
Comment (Beckyrose200) [inline app/validators/return-requirements/abstraction-period.validator.js:15]: 'When setting up a return requirement, users need to specify an abstraction period. For this, the users must input a valid `from` and `to` date. If a user does not enter valid dates the validation will return an error.'

I would reword this comment. The comment on line 15 goes over the line length and in parts of the comment it doesn't read properly.

File: app/validators/return-requirements/abstraction-period.validator.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+ * date for the licence. If there requirements are not met the validation will return an error.
+ *
+ * @param {Object} payload - The payload from the request to be validated.
+ *
+ * @returns {Object} - The result from calling Joi's schema.validate(). If any error are found the 'error: ' property will
```
Comment (Beckyrose200) [inline app/validators/return-requirements/abstraction-period.validator.js:20]: Gone over the line length

File: app/validators/return-requirements/abstraction-period.validator.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+  return parsePayload
+}
+
+function _validateAbstractionFromDate (fromDate) {
+  const schema = Joi.object({
+    entry: Joi.string()
+      .required()
+      .messages({
+        'string.empty': 'Select the from date of the abstraction period'
```
Comment (Beckyrose200) [inline app/validators/return-requirements/abstraction-period.validator.js:69]: The error message text here doesn't match up with the ticket. Unsure if this has been a change not documented on the ticket.

File: app/validators/return-requirements/abstraction-period.validator.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+  return schema.validate(fromDate, { abortEarly: true })
+}
+
+function _validateAbstractionToDate (toDate) {
+  const schema = Joi.object({
+    entry: Joi.string()
+      .required()
+      .messages({
+        'string.empty': 'Select the to date of the abstraction period'
```
Comment (Beckyrose200) [inline app/validators/return-requirements/abstraction-period.validator.js:88]: The error message text here doesn't match up with the ticket. Unsure if this has been a change not documented on the ticket.

**Review by Beckyrose200 — CHANGES_REQUESTED** [review]
Good work, just a few comments from me :)

File: test/controllers/return-requirements.controller.test.js [hunk 5]
```diff
@@ -9,6 +9,7 @@ const { describe, it, beforeEach, afterEach } = exports.lab = Lab.script()
 const { expect } = Code
 
 // Things we need to stub
+const AbstractionPointsService = require('../../app/services/return-requirements/abstraction-period.service.js')
```
[the remark this answers is not in the document]
Comment (rvsiyad) [reply test/controllers/return-requirements.controller.test.js:12]: Good catch, you are right!

File: app/validators/return-requirements/abstraction-period.validator.js (hunk 3 above)
Comment (rvsiyad) [reply app/validators/return-requirements/abstraction-period.validator.js:69]: Alan said to stick to this error message, makes more sense!

File: app/validators/return-requirements/abstraction-period.validator.js (hunk 4 above)
Comment (rvsiyad) [reply app/validators/return-requirements/abstraction-period.validator.js:88]: Same again!

File: app/presenters/return-requirements/abstraction-period.presenter.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    licenceId: session.data.licence.id,
+    licenceRef: session.data.licence.licenceRef,
+    abstractionPeriod: _licenceAbstractionPeriod(payload)
+  }
+
+  return data
+}
+
+function _licenceAbstractionPeriod (payload) {
+  // NOTE: The following checks whether a user has inputted any values for the abstraction period for the returns requirements.
```
Comment (Beckyrose200) [inline app/presenters/return-requirements/abstraction-period.presenter.js:28]: This comment goes over the allowed line length

File: app/services/return-requirements/abstraction-period.service.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+ * @module AbstractionPeriodService
+ */
+
+const AbstractionPeriodPresenter = require('../../presenters/return-requirements/abstraction-period.presenter.js')
+const SessionModel = require('../../models/session.model.js')
+
+/**
+ * Orchestrates fetching and presenting the data for `/return-requirements/{sessionId}/abstraction-period` page
+ *
+ * Supports generating the data needed for the abstraction period page in the return requirements setup journey. It fetches the
```
Comment (Beckyrose200) [inline app/services/return-requirements/abstraction-period.service.js:14]: This comment goes over the allowed line length

File: app/validators/return-requirements/abstraction-period.validator.js [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+ * error.
+ *
+ * @param {Object} payload - The payload from the request to be validated.
+ *
+ * @returns {Object} - The result from calling Joi's schema.validate(). If any errors are found the 'error: ' property
+ * will also exist, detailing what the issue is.
+ */
+function go (payload) {
+  const {
+    'fromAbstractionPeriod-day': fromDay,
```
Comment (Beckyrose200) [inline app/validators/return-requirements/abstraction-period.validator.js:26]: Having a mix of[ kebab case](https://www.theserverside.com/definition/Kebab-case) and [Camel case](https://www.techtarget.com/whatis/definition/CamelCase) doesn't look right here. From viewing one of [Fazels PR's ](https://github.com/DEFRA/water-abstraction-system/pull/646) I would say we need to stick to kebab case because of the way it has come from the page. I think this was something that was discussed during a dev call to sort for our coding conventions.

File: test/controllers/return-requirements.controller.test.js [hunk 9]
```diff
@@ -41,12 +42,18 @@ describe('Return requirements controller', () => {
   })
 
   describe('GET /return-requirements/{sessionId}/abstraction-period', () => {
+    beforeEach(async () => {
+      Sinon.stub(AbstractionPeriodService, 'go').resolves({
+        id: '8702b98f-ae51-475d-8fcc-e049af8b8d38', pageTitle: 'Enter the abstraction period for the requirements for returns'
```
Comment (Beckyrose200) [inline test/controllers/return-requirements.controller.test.js:47]: This goes over the allowed line length

**Review by Beckyrose200 — CHANGES_REQUESTED** [review]
Looks really good just a few comments :)

File: test/services/return-requirements/submit-abstraction-period.service.test.js [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+
+        expect(refreshedSession.data.abstractionPeriod).to.equal({
+          'to-abstraction-period-day': '02',
+          'from-abstraction-period-day': '01',
+          'to-abstraction-period-month': '7',
+          'from-abstraction-period-month': '12'
+        })
+      })
+
+      it('returns an empty object (no page data needed for a redirect', async () => {
```
Comment (Jozzey) [inline test/services/return-requirements/submit-abstraction-period.service.test.js:63]: Typo
```suggestion
      it('returns an empty object (no page data needed for a redirect)', async () => {
```

**Review by Jozzey — COMMENTED** [review]
Looks good to me apart from Rebecca's comments and a typo I spotted

File: test/validators/no-returns-required/abstraction-period.validator.test.js [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+const { describe, it } = exports.lab = Lab.script()
+const { expect } = Code
+
+// Thing under test
+const AbstractionPeriodValidator = require('../../../app/validators/return-requirements/abstraction-period.validator.js')
+
+describe('Abstraction Period validator', () => {
+  describe('when valid data is provided', () => {
+    const payload = {
```
[the remark this answers is not in the document]
Comment (rvsiyad) [reply test/validators/no-returns-required/abstraction-period.validator.test.js:15]: Having a issue adding let payload as a global variable, causing tests to fail. Only way I have got it to work so far is buy putting each tests payload into a beforeEach block.

```suggestion
let payload
beforeEach(() => {
   payload = {}
   })
```

Which way would be preferred?

**Review by Beckyrose200 — APPROVED** [review]
Looks great
